drv/stm32xx-i2c-server/src/main.rs

```diff
@@ -193,22 +266,34 @@ fn reset_if_needed(
     controller: &I2cController<'_>,
     port: PortIndex,
     muxes: &[I2cMux<'_>],
-    mux: Option<(Mux, Segment)>,
+    muxmap: &mut MuxMap,
 ) {
     if reset_needed(code) {
-        reset(controller, port, muxes, mux)
+        reset(controller, port, muxes, muxmap)
     }
 }
 
 include!(concat!(env!("OUT_DIR"), "/i2c_config.rs"));
 
 type PortMap = FixedMap<Controller, PortIndex, { i2c_config::NCONTROLLERS }>;
 
-type MuxMap = FixedMap<
-    (Controller, PortIndex),
-    (Mux, Segment),
-    { i2c_config::NMUXEDBUSES },
->;
+#[derive(Copy, Clone, Debug)]
+enum MuxState {
+    /// a mux+segment have been explicitly enabled
+    Enabled(Mux, Segment),
+
+    /// state is unknown: zero, one, or more mux+segment(s) may be enabled
+    Unknown,
+}
+
+///
+/// Contains the mux state on a per-bus basis.  If no mux+segment is enabled
+/// for a bus (that is, if any/all muxes on a bus have been explicitly had
+/// all segments disabled), there will not be an entry for the bus in this
+/// map.
+///
+type MuxMap =
+    FixedMap<(Controller, PortIndex), MuxState, { i2c_config::NMUXEDBUSES }>;
```

Comment:
TIL that this is our own `FixedMap` implementation (which is a linear scan), and not the [`fixed-map` crate](https://docs.rs/fixed-map/latest/fixed_map/).

Reply:
TIL about the the [`fixed-map` crate](https://docs.rs/fixed-map/latest/fixed_map/). 😉 

It looks like the `no-std` approach there is [relatively recent](https://github.com/udoprog/fixed-map/commit/6237fd7fbf3aa7728796f64c34fb38c5cfb9d843) (and certainly postdates our use), so it might be worth exploring.  That said, we expect these maps to be nanoscopic -- `i2c_config::NMUXEDBUSES` is 2 on Gimlet, 2 on Sidecar, 1 on Donglet and 0 on the PSC -- so it's not clear how much upside there is to be had.

Reply:
Agreed, I'll keep my "accidentally quadratic" flag stowed while `N` is so small.